File: trading_api/go_analysis/utils/indicators.go
package utils

import (
	"math"
)

// CalculateEMA calculates Exponential Moving Average
func CalculateEMA(prices []float64, period int) []float64 {
	if len(prices) < period {
		return make([]float64, len(prices))
	}

	result := make([]float64, len(prices))
	multiplier := 2.0 / float64(period+1)

	// Calculate initial SMA
	sum := 0.0
	for i := 0; i < period; i++ {
		sum += prices[i]
	}
	result[period-1] = sum / float64(period)

	// Calculate EMA
	for i := period; i < len(prices); i++ {
		result[i] = (prices[i]-result[i-1])*multiplier + result[i-1]
	}

	return result
}

// CalculateRSI calculates Relative Strength Index
func CalculateRSI(prices []float64, period int) []float64 {
	if len(prices) < period+1 {
		return make([]float64, len(prices))
	}

	result := make([]float64, len(prices))
	gains := make([]float64, len(prices))
	losses := make([]float64, len(prices))

	// Calculate price changes
	for i := 1; i < len(prices); i++ {
		change := prices[i] - prices[i-1]
		if change > 0 {
			gains[i] = change
		} else {
			losses[i] = -change
		}
	}

	// Calculate initial average gain/loss
	avgGain := 0.0
	avgLoss := 0.0
	for i := 1; i <= period; i++ {
		avgGain += gains[i]
		avgLoss += losses[i]
	}
	avgGain /= float64(period)
	avgLoss /= float64(period)

	// Calculate RSI
	for i := period; i < len(prices); i++ {
		if i > period {
			avgGain = (avgGain*float64(period-1) + gains[i]) / float64(period)
			avgLoss = (avgLoss*float64(period-1) + losses[i]) / float64(period)
		}

		if avgLoss == 0 {
			result[i] = 100
		} else {
			rs := avgGain / avgLoss
			result[i] = 100 - (100 / (1 + rs))
		}
	}

	return result
}

// CalculateATR calculates Average True Range
func CalculateATR(high, low, close []float64, period int) []float64 {
	if len(high) < period+1 {
		return make([]float64, len(high))
	}

	result := make([]float64, len(high))
	trueRanges := make([]float64, len(high))

	// Calculate True Range
	for i := 1; i < len(high); i++ {
		highLow := high[i] - low[i]
		highClose := math.Abs(high[i] - close[i-1])
		lowClose := math.Abs(low[i] - close[i-1])

		trueRanges[i] = math.Max(highLow, math.Max(highClose, lowClose))
	}

	// Calculate initial ATR (SMA of TR)
	sum := 0.0
	for i := 1; i <= period; i++ {
		sum += trueRanges[i]
	}
	result[period] = sum / float64(period)

	// Calculate ATR (EMA of TR)
	for i := period + 1; i < len(high); i++ {
		result[i] = (result[i-1]*float64(period-1) + trueRanges[i]) / float64(period)
	}

	return result
}

File: trading_api/rust_analysis/src/indicators.rs
/// Calculate Exponential Moving Average (EMA)
pub fn calculate_ema(prices: &[f64], period: usize) -> Vec<f64> {
    let len = prices.len();
    
    if len < period {
        return vec![0.0; len];
    }
    
    let mut ema = vec![0.0; len];
    let multiplier = 2.0 / (period as f64 + 1.0);
    
    // Initial SMA
    let sum: f64 = prices[..period].iter().sum();
    ema[period - 1] = sum / period as f64;
    
    // Calculate EMA
    for i in period..len {
        ema[i] = (prices[i] - ema[i - 1]) * multiplier + ema[i - 1];
    }
    
    ema
}

/// Calculate Relative Strength Index (RSI)
pub fn calculate_rsi(prices: &[f64], period: usize) -> Vec<f64> {
    let len = prices.len();
    
    if len < period + 1 {
        return vec![0.0; len];
    }
    
    let mut rsi = vec![0.0; len];
    let mut gains = vec![0.0; len];
    let mut losses = vec![0.0; len];
    
    // Calculate price changes
    for i in 1..len {
        let change = prices[i] - prices[i - 1];
        if change > 0.0 {
            gains[i] = change;
        } else {
            losses[i] = -change;
        }
    }
    
    // Initial average gain/loss
    let mut avg_gain: f64 = gains[1..=period].iter().sum::<f64>() / period as f64;
    let mut avg_loss: f64 = losses[1..=period].iter().sum::<f64>() / period as f64;
    
    // Calculate RSI using Wilder's smoothing
    let alpha = 1.0 / period as f64;
    
    for i in period..len {
        avg_gain = avg_gain * (1.0 - alpha) + gains[i] * alpha;
        avg_loss = avg_loss * (1.0 - alpha) + losses[i] * alpha;
        
        if avg_loss == 0.0 {
            rsi[i] = 100.0;
        } else {
            let rs = avg_gain / avg_loss;
            rsi[i] = 100.0 - (100.0 / (1.0 + rs));
        }
    }
    
    rsi
}

/// Calculate Average True Range (ATR)
pub fn calculate_atr(high: &[f64], low: &[f64], close: &[f64], period: usize) -> Vec<f64> {
    let len = high.len();
    
    if len < period + 1 {
        return vec![0.0; len];
    }
    
    let mut atr = vec![0.0; len];
    let mut true_ranges = vec![0.0; len];
    
    // Calculate True Range
    for i in 1..len {
        let hl = high[i] - low[i];
        let hc = (high[i] - close[i - 1]).abs();
        let lc = (low[i] - close[i - 1]).abs();
        true_ranges[i] = hl.max(hc).max(lc);
    }
    
    // Initial ATR (SMA of TR)
    let sum: f64 = true_ranges[1..=period].iter().sum();
    atr[period] = sum / period as f64;
    
    // Wilder's smoothing
    let alpha = 1.0 / period as f64;
    for i in (period + 1)..len {
        atr[i] = atr[i - 1] * (1.0 - alpha) + true_ranges[i] * alpha;
    }
    
    atr
}
